CRM/Core/Resources.php

```diff
@@ -272,21 +275,27 @@ public function addVars($nameSpace, $vars) {
    * Extensions and components should generally use addVars instead.
    *
    * @param array $settings
+   * @param string $region
+   *   The region to add settings to (eg. for payment processors usually billing-block)
+   *
    * @return CRM_Core_Resources
    */
-  public function addSetting($settings) {
-    $this->settings = $this->mergeSettings($this->settings, $settings);
-    if (!$this->addedSettings) {
+  public function addSetting($settings, $region = NULL) {
+    if (!$region) {
       $region = self::isAjaxMode() ? 'ajax-snippet' : 'html-header';
-      $resources = $this;
-      CRM_Core_Region::instance($region)->add([
-        'callback' => function (&$snippet, &$html) use ($resources) {
-          $html .= "\n" . $resources->renderSetting();
-        },
-        'weight' => -100000,
-      ]);
-      $this->addedSettings = TRUE;
```

Comment:
Can you explain a little more what you did with this variable `addedSettings` and why it's not needed anymore?

Reply:
`addedSettings` gets set to TRUE the first time `addSettings()` is called. If settings are being added to the same region that is ok but if you are adding settings to a different region they will never be added.  By doing the add/update we ensure they are not added multiple times to the same region but *possibly* increase processing overhead slightly as an update will regenerate the region (via `renderSetting()`). The actual settings are settings are added whether or not you re-render the region - so if you call to add settingsgroup1 to ajax-snippet and then again to add settingsgroup2 to ajax-snippet you will always end up with both in the final render.
**But** if you want to add settings to a *different* region they would never get added if we had the boolean `addedSettings` because it would be set to TRUE after the first region.
A good example is `coreForm` settings which are added to ajax-snippet here: https://github.com/civicrm/civicrm-core/blob/master/CRM/Contribute/Form/AbstractEditPayment.php#L238 in the same request that reloads the billing-block and Stripe needs to insert settings into the billing-block region. *We probably don't actually need the `coreForm` adding to ajax-snippet in this subsequent call but it doesn't do any harm - it doesn't get rendered.

Reply:
@colemanw I hope that's not too long and rambling!